File: python-backend/app/api/routes/tracking.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.api.dependencies import get_session, get_tracking_service
from app.infra.persistence.session_store import Session
from app.models.enums import TrackingDirection
from app.services.tracking_service import TrackingService

router = APIRouter(prefix="/tracking", tags=["Tracking"])


class InitializeRequest(BaseModel):
    frame_index: int
    roi: list[int]  # [x, y, w, h]


class PropagateRequest(BaseModel):
    direction: TrackingDirection = TrackingDirection.FORWARD
    max_frames: int | None = None
    auto_segment: bool = False


@router.post("/initialize")
async def initialize_tracking(
    request: InitializeRequest,
    session: Session = Depends(get_session),
    service: TrackingService = Depends(get_tracking_service),
):
    if len(request.roi) != 4:
        raise HTTPException(
            400,
            detail={
                "code": "INVALID_ROI",
                "message": "ROI must be [x, y, w, h]",
            },
        )
    roi = (request.roi[0], request.roi[1], request.roi[2], request.roi[3])
    try:
        return service.initialize(session, request.frame_index, roi)
    except (ValueError, IndexError) as exc:
        raise HTTPException(400, detail={"code": "INIT_FAILED", "message": str(exc)})


@router.post("/propagate")
async def propagate_tracking(
    request: PropagateRequest,
    session: Session = Depends(get_session),
    service: TrackingService = Depends(get_tracking_service),
):
    try:
        return service.propagate(
            session,
            direction=request.direction,
            max_frames=request.max_frames,
            auto_segment=request.auto_segment,
        )
    except ValueError as exc:
        raise HTTPException(400, detail={"code": "PROPAGATE_FAILED", "message": str(exc)})


@router.get("/state")
async def get_tracking_state(
    session: Session = Depends(get_session),
    service: TrackingService = Depends(get_tracking_service),
):
    return service.get_state(session)


@router.delete("/clear")
async def clear_tracking(
    session: Session = Depends(get_session),
    service: TrackingService = Depends(get_tracking_service),
):
    return service.clear(session)

File: src/lib/api/tracking.ts
import { apiClient } from '@/lib/api/client';

export interface TrackingInitResponse {
  frame_index: number;
  roi: [number, number, number, number];
  status: string;
}

export interface TrackingFrameResult {
  frame_index: number;
  bbox: [number, number, number, number];
  confidence: number;
  success: boolean;
}

export interface TrackingPropagateResponse {
  direction: 'forward' | 'backward';
  tracked_frames: number;
  total_frames: number;
  results: TrackingFrameResult[];
}

export interface TrackingStateResponse {
  status: string;
  start_frame?: number | null;
  roi?: [number, number, number, number] | null;
  results: TrackingFrameResult[];
  num_tracked?: number;
}

export async function initializeTracking(
  frameIndex: number,
  roi: [number, number, number, number],
): Promise<TrackingInitResponse> {
  const response = await apiClient.post<TrackingInitResponse>('/tracking/initialize', {
    frame_index: frameIndex,
    roi,
  });
  return response.data;
}

export async function propagateTracking(
  direction: 'forward' | 'backward',
  maxFrames?: number,
  autoSegment?: boolean,
): Promise<TrackingPropagateResponse> {
  const response = await apiClient.post<TrackingPropagateResponse>('/tracking/propagate', {
    direction,
    max_frames: maxFrames,
    auto_segment: autoSegment,
  });
  return response.data;
}

export async function getTrackingState(): Promise<TrackingStateResponse> {
  const response = await apiClient.get<TrackingStateResponse>('/tracking/state');
  return response.data;
}

export async function clearTracking(): Promise<void> {
  await apiClient.post('/tracking/clear');
}
